File: android/src/androidTest/java/org/ligi/gobandroid_hd/uitest/TheEditGameActivity.java
package org.ligi.gobandroid_hd.uitest;

import android.test.suitebuilder.annotation.MediumTest;

import com.squareup.spoon.Spoon;

import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.R;
import org.ligi.gobandroid_hd.TestApp;
import org.ligi.gobandroid_hd.etc.AppModule;
import org.ligi.gobandroid_hd.logic.Cell;
import org.ligi.gobandroid_hd.logic.CellFactory;
import org.ligi.gobandroid_hd.logic.CellImpl;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.gobandroid_hd.logic.markers.SquareMarker;
import org.ligi.gobandroid_hd.logic.markers.TextMarker;
import org.ligi.gobandroid_hd.logic.markers.TriangleMarker;
import org.ligi.gobandroid_hd.model.GameProvider;
import org.ligi.gobandroid_hd.ui.editing.EditGameActivity;
import org.ligi.gobandroid_hd.base.BaseIntegration;

import javax.inject.Inject;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withContentDescription;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static org.assertj.core.api.Assertions.assertThat;
import static org.ligi.gobandroid_hd.base.GoViewActions.tapStone;

public class TheEditGameActivity extends BaseIntegration<EditGameActivity> {

    @Inject
    GameProvider gameProvider;

    public TheEditGameActivity() {
        super(EditGameActivity.class);
    }

    @Override
    public void setUp() throws Exception {
        super.setUp();
        TestApp.component().inject(this);
    }

    @MediumTest
    public void testThatGoBoardIsThere() {
        final EditGameActivity activity = getActivity();

        onView(withId(R.id.go_board)).check(matches(isDisplayed()));
        Spoon.screenshot(activity, "edit");
    }

    @MediumTest
    public void testThatLettersWork() {
        gameProvider.set(new GoGame(9));

        final EditGameActivity activity = getActivity();

        onView(withContentDescription(getString(R.string.letter))).perform(click());

        tap9x3Field();

        Spoon.screenshot(activity, "letters");
    }

    private void tap9x3Field() {
        for (Cell cell : CellFactory.getAllCellsForRect(9, 3)) {
            onView(withId(R.id.go_board)).perform(tapStone(cell, gameProvider.get()));
        }
    }

    @MediumTest
    public void testThatNumbersWork() {
        gameProvider.set(new GoGame(9));

        final EditGameActivity activity = getActivity();

        onView(withContentDescription(getString(R.string.number))).perform(click());

        tap9x3Field();

        for (Cell cell : CellFactory.getAllCellsForRect(9, 3)) {
            assertThat(gameProvider.get().getActMove().getMarkers()).contains(new TextMarker(cell, "" + (9 * cell.getY() + cell.getX() + 1)));
        }

        Spoon.screenshot(activity, "numbers");
    }


    @MediumTest
    public void testThatSquareWorks() {
        gameProvider.set(new GoGame(9));

        final EditGameActivity activity = getActivity();

        onView(withContentDescription(getString(R.string.square))).perform(click());

        final CellImpl cell = new CellImpl(1, 2);
        onView(withId(R.id.go_board)).perform(tapStone(cell, gameProvider.get()));

        assertThat(gameProvider.get().getActMove().getMarkers()).contains(new SquareMarker(cell));

        Spoon.screenshot(activity, "square");
    }


    @MediumTest
    public void testThatTriangleWorks() {
        gameProvider.set(new GoGame(9));

        final EditGameActivity activity = getActivity();

        onView(withContentDescription(getString(R.string.triangle))).perform(click());

        onView(withId(R.id.go_board)).perform(tapStone(new CellImpl(1, 2), gameProvider.get()));

        assertThat(gameProvider.get().getActMove().getMarkers()).contains(new TriangleMarker(new CellImpl(1, 2)));

        Spoon.screenshot(activity, "triangle");
    }
}

File: android/src/androidTest/java/org/ligi/gobandroid_hd/uitest/TheEditGameActivity.kt
package org.ligi.gobandroid_hd.uitest

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.ViewActions.click
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers.*
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.lazy
import com.squareup.spoon.Spoon
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.ligi.gobandroid_hd.App
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.logic.CellImpl
import org.ligi.gobandroid_hd.logic.GoGame
import org.ligi.gobandroid_hd.logic.markers.GoMarker
import org.ligi.gobandroid_hd.logic.markers.SquareMarker
import org.ligi.gobandroid_hd.logic.markers.TextMarker
import org.ligi.gobandroid_hd.logic.markers.TriangleMarker
import org.ligi.gobandroid_hd.model.GameProvider
import org.ligi.gobandroid_hd.test_helper_functions.getAllCellsForRect
import org.ligi.gobandroid_hd.test_helper_functions.tapStone
import org.ligi.gobandroid_hd.ui.editing.EditGameActivity
import org.ligi.trulesk.TruleskActivityRule

class TheEditGameActivity {

    @get:Rule
    val rule = TruleskActivityRule(EditGameActivity::class.java, false)

    val gameProvider: GameProvider by App.kodein.lazy.instance()

    @Test
    fun testThatGoBoardIsThere() {
        rule.launchActivity(null)

        onView(withId(R.id.go_board)).check(matches(isDisplayed()))
        rule.screenShot("edit")
    }

    @Test
    fun testThatLettersWork() {
        gameProvider.set(GoGame(9))

        val activity = rule.launchActivity(null)

        onView(withContentDescription(activity.getString(R.string.letter))).perform(click())

        tap9x3Field()

        Spoon.screenshot(activity, "letters")
    }

    private fun tap9x3Field() {
        for (cell in getAllCellsForRect(9, 3)) {
            onView(withId(R.id.go_board)).perform(tapStone(cell, gameProvider.get()))
        }
    }

    @Test
    fun testThatNumbersWork() {
        gameProvider.set(GoGame(9))

        val activity = rule.launchActivity(null)

        onView(withContentDescription(activity.getString(R.string.number))).perform(click())

        tap9x3Field()

        rule.screenShot("numbers")

        for (cell in getAllCellsForRect(9, 3)) {
            assertThat(gameProvider.get().actMove.markers).contains(TextMarker(cell, "" + (3 * cell.y + cell.x + 1)))
        }
    }


    @Test
    fun testThatSquareWorks() {
        gameProvider.set(GoGame(9))

        val activity = rule.launchActivity(null)

        onView(withContentDescription(activity.getString(R.string.square))).perform(click())

        val cell = CellImpl(1, 2)
        onView(withId(R.id.go_board)).perform(tapStone(cell, gameProvider.get()))

        assertThat<GoMarker>(gameProvider.get().actMove.markers).contains(SquareMarker(cell))

        rule.screenShot("square")
    }


    @Test
    fun testThatTriangleWorks() {
        gameProvider.set(GoGame(9))

        val activity = rule.launchActivity(null)

        onView(withContentDescription(activity.getString(R.string.triangle))).perform(click())

        onView(withId(R.id.go_board)).perform(tapStone(CellImpl(1, 2), gameProvider.get()))

        assertThat<GoMarker>(gameProvider.get().actMove.markers).contains(TriangleMarker(CellImpl(1, 2)))

        rule.screenShot("triangle")
    }
}
